packages/utils/plugins-helpers/src/federation.ts

```diff
@@ -28,14 +30,67 @@ export const federationSpec = parse(/* GraphQL */ `
   directive @key(fields: _FieldSet!) on OBJECT | INTERFACE
 `);
 
+interface TypeMeta {
+  hasResolveReference: boolean;
+  resolvableKeyDirectives: readonly DirectiveNode[];
+}
+
+export type FederationMeta = { [typeName: string]: TypeMeta };
+
 /**
- * Adds `__resolveReference` in each ObjectType involved in Federation.
+ * Adds `__resolveReference` in each ObjectType and InterfaceType involved in Federation.
+ * We do this to utilise the existing FieldDefinition logic of the plugin, which includes many logic:
+ * - mapper
+ * - return type
  * @param schema
  */
-export function addFederationReferencesToSchema(schema: GraphQLSchema): GraphQLSchema {
-  return mapSchema(schema, {
+export function addFederationReferencesToSchema(schema: GraphQLSchema): {
```

Comment:
We now collect Federation meta in this function because this is called before calling visitor.
This means we no longer have to collect meta as we find entities during visitor -> simpler, more predictable, and more performant